README.md
# weixin-openapi

[![npm][npm-img]][npm-url]
[![build status][build-img]][build-img]
[![npm download][download-img]][download-url]

[npm-img]: https://img.shields.io/npm/v/@codit-run/weixin-openapi.svg
[npm-url]: https://www.npmjs.org/package/@codit-run/weixin-openapi
[build-img]: https://github.com/codit-run/weixin-openapi/actions/workflows/ci.yml/badge.svg
[build-url]: https://github.com/codit-run/weixin-openapi/actions/workflows/ci.yml
[download-img]: https://img.shields.io/npm/dm/@codit-run/weixin-openapi.svg
[download-url]: https://www.npmjs.org/package/@codit-run/weixin-openapi

Weixin OpenAPI

## Installation

```js
$ npm install @codit-run/weixin-openapi
```

## Open APIs

- Mini Program

## TODO
- Monorepo
- Documentation
- Do real requests in tests


CHANGELOG.md
## [0.1.0] (2023-07-24)

### Initial commit


src/core/error.spec.ts

import { describe, expect, it } from 'vitest'
import { createError } from './error.js'

describe('createError', () => {
  it('creates an "WeixinAuthCodeError"', () => {
    const authCodeError = createError('Unable to create a session.', {
      errcode: 40029,
      errmsg: 'invalid code',
    })
    const authcodeErrorJson = {
      name: 'WeixinAuthCodeError',
      message: 'Unable to create a session: invalid code.',
      response: {
        errcode: 40029,
        errmsg: 'invalid code',
      },
    }

    expect(authCodeError.toJSON()).toStrictEqual(authcodeErrorJson)
    expect(JSON.stringify(authCodeError)).toBe(JSON.stringify(authcodeErrorJson))
    expect(authCodeError.toString()).toBe('WeixinAuthCodeError: Unable to create a session: invalid code.') // default is `${name}: ${message}`
  })

  it('creates a general "WeixinError"', () => {
    const generalError = createError('Unable to create a session.', {
      errcode: 40013,
      errmsg: 'invalid appid',
    })
    const generalErrorJson = {
      name: 'WeixinError',
      message: 'Unable to create a session.',
      response: {
        errcode: 40013,
        errmsg: 'invalid appid',
      },
    }

    expect(generalError.toJSON()).toStrictEqual(generalErrorJson)
  })
})


src/core/client.spec.ts
import { beforeEach, describe, expect, it, test, vi } from 'vitest'

import { WeixinClient } from './client.js'
import * as fetcher from './fetcher.js'
import { WeixinMemoryStore } from './memory-store.js'

vi.mock('./fetcher.js', async (): Promise<typeof fetcher> => {
  return {
    request: vi.fn(),
    createAccessToken: vi.fn(async () => ({
      access_token: 'test_access_token',
      expires_in: 7200,
    })),
    createStableAccessToken: vi.fn(async () => ({
      access_token: 'test_access_token',
      expires_in: 7200,
    })),
    exchangeCodeForSession: vi.fn(async () => ({
      session_key: 'test_session_key',
      openid: 'test_openid',
    })),
    exchangeCodeForToken: vi.fn(async () => ({
      access_token: 'test_access_token',
      expires_in: 7200,
      refresh_token: 'test_refresh_token',
      openid: 'test_openid',
      scope: 'snsapi_login',
      unionid: 'test_unionid',
    })),
  }
})

const store = new WeixinMemoryStore()

const client = new WeixinClient({
  appid: 'appid',
  secret: 'secret',
  store,
})

const successData = {
  quota: {
    daily_limit: 100000000,
    used: 7,
    remain: 99999993,
  },
}

const invalidAccessTokenData = {
  errcode: 40001,
  errmsg: 'invalid credential',
}

const quotaExeededData = {
  errcode: 45009,
  errmsg: 'daily quota limit',
}

function createResponse(data: object, raw: boolean): Response {
  return {
    headers: {
      get(name) {
        if (raw && !('errcode' in data && data.errcode !== 0))
          return 'image/jpeg'
        return 'application/json'
      },
    },
    async json() { return data },
  } as Response
}

beforeEach(async () => {
  await store._clear()
  return () => vi.restoreAllMocks()
})

describe('#get', () => {
  it('returns success response', async () => {
    vi.mocked(fetcher.request).mockResolvedValue(createResponse(successData, false))

    const res = await client.get('/some/path', {})
    expect(res).toStrictEqual(successData)
    expect(fetcher.createAccessToken).toBeCalledTimes(1)
  })

  it('returns success response after one access token auto-retry', async () => {
    vi.mocked(fetcher.request)
      .mockResolvedValueOnce(createResponse(invalidAccessTokenData, false))
      .mockResolvedValueOnce(createResponse(successData, false))

    const res = await client.get('/some/path', {})
    expect(res).toStrictEqual(successData)
    expect(fetcher.createAccessToken).toBeCalledTimes(2)
  })

  it('throws an error after one access token auto-retry', async () => {
    vi.mocked(fetcher.request).mockResolvedValue(createResponse(invalidAccessTokenData, false))

    await expect(client.get('/some/path', {})).rejects.toThrow(/^Request to '.+' is failed.$/)
    expect(fetcher.createAccessToken).toBeCalledTimes(2)
  })

  it('throws an error without access token auto-retry', async () => {
    vi.mocked(fetcher.request).mockResolvedValue(createResponse(quotaExeededData, false))

    await expect(client.get('/some/path', {})).rejects.toThrow(/^Request to '.+' is failed.$/)
    expect(fetcher.createAccessToken).toBeCalledTimes(1)
  })
})

describe('#post', () => {
  it('returns success response', async () => {
    vi.mocked(fetcher.request).mockResolvedValue(createResponse(successData, false))

    const res = await client.post('/some/path', {})
    expect(res).toStrictEqual(successData)
    expect(fetcher.createAccessToken).toBeCalledTimes(1)
  })

  it('returns success response after one access token auto-retry', async () => {
    vi.mocked(fetcher.request)
      .mockResolvedValueOnce(createResponse(invalidAccessTokenData, false))
      .mockResolvedValueOnce(createResponse(successData, false))

    const res = await client.post('/some/path', {})
    expect(res).toStrictEqual(successData)
    expect(fetcher.createAccessToken).toBeCalledTimes(2)
  })

  it('throws an error after one access token auto-retry', async () => {
    vi.mocked(fetcher.request).mockResolvedValue(createResponse(invalidAccessTokenData, false))

    await expect(client.post('/some/path', {})).rejects.toThrow(/^Request to '.+' is failed.$/)
    expect(fetcher.createAccessToken).toBeCalledTimes(2)
  })

  it('throws an error without access token auto-retry', async () => {
    vi.mocked(fetcher.request).mockResolvedValue(createResponse(quotaExeededData, false))

    await expect(client.post('/some/path', {})).rejects.toThrow(/^Request to '.+' is failed.$/)
    expect(fetcher.createAccessToken).toBeCalledTimes(1)
  })
})

describe('#getRaw', () => {
  it('returns success response', async () => {
    vi.mocked(fetcher.request).mockResolvedValue(createResponse(successData, true))

    const res = await client.getRaw('/some/path', {})
    expect(await res.json()).toStrictEqual(successData)
    expect(fetcher.createAccessToken).toBeCalledTimes(1)
  })

  it('returns success response after one access token auto-retry', async () => {
    vi.mocked(fetcher.request)
      .mockResolvedValueOnce(createResponse(invalidAccessTokenData, true))
      .mockResolvedValueOnce(createResponse(successData, true))

    const res = await client.getRaw('/some/path', {})
    expect(await res.json()).toStrictEqual(successData)
    expect(fetcher.createAccessToken).toBeCalledTimes(2)
  })

  it('throws an error after one access token auto-retry', async () => {
    vi.mocked(fetcher.request).mockResolvedValue(createResponse(invalidAccessTokenData, true))

    await expect(client.getRaw('/some/path', {})).rejects.toThrow(/^Request to '.+' is failed.$/)
    expect(fetcher.createAccessToken).toBeCalledTimes(2)
  })

  it('throws an error without access token auto-retry', async () => {
    vi.mocked(fetcher.request).mockResolvedValue(createResponse(quotaExeededData, true))

    await expect(client.getRaw('/some/path', {})).rejects.toThrow(/^Request to '.+' is failed.$/)
    expect(fetcher.createAccessToken).toBeCalledTimes(1)
  })
})

describe('#postRaw', () => {
  it('returns success response', async () => {
    vi.mocked(fetcher.request).mockResolvedValue(createResponse(successData, true))

    const res = await client.postRaw('/some/path', {})
    expect(await res.json()).toStrictEqual(successData)
    expect(fetcher.createAccessToken).toBeCalledTimes(1)
  })

  it('returns success response after one access token auto-retry', async () => {
    vi.mocked(fetcher.request)
      .mockResolvedValueOnce(createResponse(invalidAccessTokenData, true))
      .mockResolvedValueOnce(createResponse(successData, true))

    const res = await client.postRaw('/some/path', {})
    expect(await res.json()).toStrictEqual(successData)
    expect(fetcher.createAccessToken).toBeCalledTimes(2)
  })

  it('throws an error after one access token auto-retry', async () => {
    vi.mocked(fetcher.request).mockResolvedValue(createResponse(invalidAccessTokenData, true))

    await expect(client.postRaw('/some/path', {})).rejects.toThrow(/^Request to '.+' is failed.$/)
    expect(fetcher.createAccessToken).toBeCalledTimes(2)
  })

  it('throws an error without access token auto-retry', async () => {
    vi.mocked(fetcher.request).mockResolvedValue(createResponse(quotaExeededData, true))

    await expect(client.postRaw('/some/path', {})).rejects.toThrow(/^Request to '.+' is failed.$/)
    expect(fetcher.createAccessToken).toBeCalledTimes(1)
  })
})

describe('#getAccessToken', () => {
  it('returns a token', async () => {
    vi.mocked(fetcher.createAccessToken)
      .mockResolvedValueOnce({ access_token: 'test_access_token1', expires_in: 7200 })
      .mockResolvedValueOnce({ access_token: 'test_access_token2', expires_in: 7200 })
      .mockResolvedValueOnce({ access_token: 'test_access_token3', expires_in: 7200 })

    const accessToken = await client.getAccessToken()

    expect(accessToken).toBe('test_access_token1')
    expect(fetcher.createAccessToken).toBeCalledTimes(1)

    const accessToken2 = await client.getAccessToken() // get again
    expect(accessToken2).toBe(accessToken) // not changed, cache hits
    expect(fetcher.createAccessToken).toBeCalledTimes(1)

    const accessToken3 = await client.getAccessToken(true) // force refresh
    expect(accessToken3).toBe('test_access_token2')
    expect(fetcher.createAccessToken).toBeCalledTimes(2)
  })

  it('throws an error', async () => {
    vi.mocked(fetcher.createAccessToken).mockResolvedValueOnce({
      errcode: 40125,
      errmsg: 'invalid appsecret rid: 64b79aff-477f027c-06415867',
    })
    await expect(client.getAccessToken()).rejects.toThrow('Unable to get an access token.')
    expect(fetcher.createAccessToken).toBeCalledTimes(1)
  })
})

describe('#getStableAccessToken', () => {
  it('returns a token', async () => {
    vi.mocked(fetcher.createStableAccessToken)
      .mockResolvedValueOnce({ access_token: 'test_access_token1', expires_in: 7200 })
      .mockResolvedValueOnce({ access_token: 'test_access_token2', expires_in: 7200 })
      .mockResolvedValueOnce({ access_token: 'test_access_token3', expires_in: 7200 })

    const accessToken = await client.getStableAccessToken()

    expect(accessToken).toBe('test_access_token1')
    expect(fetcher.createStableAccessToken).toBeCalledTimes(1)

    const accessToken2 = await client.getStableAccessToken() // get again
    expect(accessToken2).toBe(accessToken) // not changed, cache hits
    expect(fetcher.createStableAccessToken).toBeCalledTimes(1)

    // Limit: 20/day quota and 1/30s rate limit
    const accessToken3 = await client.getStableAccessToken(true)
    expect(accessToken3).toBe('test_access_token2')
    expect(fetcher.createStableAccessToken).toBeCalledTimes(2)
  })

  it('throws an error', async () => {
    vi.mocked(fetcher.createStableAccessToken).mockResolvedValueOnce({
      errcode: 45009,
      errmsg: 'reach max api daily quota limit rid: 64b279b3-2eddb111-78605a3e',
    })
    await expect(client.getStableAccessToken()).rejects.toThrow('Unable to get a stable access token.')
    expect(fetcher.createStableAccessToken).toBeCalledTimes(1)
  })
})

describe('#createSessionFromCode', () => {
  it('returns session', async () => {
    const session = await client.createSessionFromCode('authcode')

    expect(session).toStrictEqual({
      session_key: 'test_session_key',
      openid: 'test_openid',
    })
    expect(fetcher.exchangeCodeForSession).toBeCalledTimes(1)
  })

  it('throws an error', async () => {
    vi.mocked(fetcher.exchangeCodeForSession).mockResolvedValueOnce({
      errcode: 40029,
      errmsg: 'invalid code, rid: 64b797fc-1d6a3114-6740d76e',
    })
    await expect(client.createSessionFromCode('authcode')).rejects.toThrow('Unable to create a session from an authorization code: invalid code.')
    expect(fetcher.exchangeCodeForSession).toBeCalledTimes(1)
  })
})

describe('#createTokenFromCode', () => {
  it('returns an access token', async () => {
    const token = await client.createTokenFromCode('authcode')

    expect(token).toStrictEqual({
      access_token: 'test_access_token',
      expires_in: 7200,
      refresh_token: 'test_refresh_token',
      openid: 'test_openid',
      scope: 'snsapi_login',
      unionid: 'test_unionid',
    })
    expect(fetcher.exchangeCodeForToken).toBeCalledTimes(1)
  })

  it('throws an error', async () => {
    vi.mocked(fetcher.exchangeCodeForToken).mockResolvedValueOnce({
      errcode: 40163,
      errmsg: 'code been used, rid: 64ba6749-265e5bf9-42ca80e9',
    })
    await expect(client.createTokenFromCode('authcode')).rejects.toThrow('Unable to create a token from an authorization code: code has been used.')
    expect(fetcher.exchangeCodeForToken).toBeCalledTimes(1)
  })
})

test('#getAPIQuota', async () => {
  const spy = vi.spyOn(client, 'post').mockResolvedValueOnce({
    quota: {
      daily_limit: 100000000,
      used: 7,
      remain: 99999993,
    },
  })
  const quota = await client.getAPIQuota('authcode')

  expect(quota).toStrictEqual({
    daily_limit: 100000000,
    used: 7,
    remain: 99999993,
  })
  expect(spy).toHaveBeenCalledTimes(1)
})


src/core/memory-store.spec.ts
import { test, expect } from 'vitest'
import { randomBytes } from 'node:crypto'
import { setTimeout as sleep } from 'node:timers/promises'
import { WeixinMemoryStore } from './memory-store.js'

const store = new WeixinMemoryStore()

async function createEntry(ttl = 1) {
  const { key, token } = randomKeyAndToken()
  await store.set(key, token, ttl)
  return { key, token }
}

function randomKeyAndToken() {
  const key = randomBytes(16).toString('base64url')
  const access_token = randomBytes(16).toString('base64url')
  return {
    key,
    token: { access_token },
  }
}

test('sets token', async () => {
  const { key, token } = randomKeyAndToken()
  expect(await store.set(key, token, 10)).toBeUndefined()
  expect(await store.get(key)).toStrictEqual(token)
})

test('gets nonexistent', async () => {
  expect(await store.get('nonexistent')).toBeNull()
})

test('gets existing', async () => {
  const { key, token } = await createEntry()
  expect(await store.get(key)).toStrictEqual(token)
})

test('gets expired', async () => {
  const { key, token } = await createEntry(1)

  expect(await store.get(key)).toStrictEqual(token)
  await sleep(1000)
  expect(await store.get(key)).toBeNull()
})

test('deletes nonexistent', async () => {
  expect(await store.delete('nonexistent')).toBeUndefined()
  expect(await store.get('nonexistent')).toBeNull()
})

test('deletes existing', async () => {
  const { key } = await createEntry()

  expect(await store.delete(key)).toBeUndefined()
  expect(await store.get(key)).toBeNull()
})


src/core/error.ts
/**
 * - [Common error codes](https://developers.weixin.qq.com/doc/oplatform/Return_codes/Return_code_descriptions_new.html)
 */

import { WeixinErrorResponse } from './types.js'

export function createError(message: string, response: WeixinErrorResponse) {
  const authCodeMessage = authcodeErrorMap(response.errcode)
  if (authCodeMessage)
    return new WeixinAuthCodeError(formatMessage(message, authCodeMessage), response)

  return new WeixinError(message, response)
}

function authcodeErrorMap(errcode: number) {
  switch (errcode) {
    case 40029: return 'invalid code'
    case 40163: return 'code has been used'
    case 42003: return 'code has expired'
  }
}

function formatMessage(title: string, detail: string): string {
  if (title.at(-1) === '.')
    return title.slice(0, -1) + ': ' + detail + '.'
  return title + ': ' + detail
}

export class WeixinError extends Error {
  readonly response: WeixinErrorResponse

  constructor(message: string, response: WeixinErrorResponse) {
    super(message, { cause: response })
    Error.captureStackTrace(this, this.constructor)
    this.name = this.constructor.name
    this.response = response
  }

  toJSON() {
    return {
      name: this.name,
      message: this.message,
      response: this.response,
    }
  }
}

// Some concret errors caused by client input.

export class WeixinAuthCodeError extends WeixinError { }


src/core/types.ts
export interface WeixinOptions {
  appid: string
  secret: string
  store: WeixinTokenStore
}

export interface WeixinErrorResponse {
  errcode: number
  errmsg: string
}

export interface WeixinAccessTokenResponse {
  access_token: string
  expires_in: number // 2 hours in seconds
}

export interface WeixinCodeSessionResponse {
  session_key: string
  unionid?: string
  openid: string
}

// TODO: To refresh access token by using `refresh_token` or fallback to the standard OAuth flow.
// https://developers.weixin.qq.com/doc/oplatform/Website_App/WeChat_Login/Wechat_Login.html
// https://learn.microsoft.com/en-us/linkedin/shared/authentication/programmatic-refresh-tokens
export interface WeixinCodeTokenResponse {
  access_token: string  // 2 hours in seconds
  refresh_token: string // 30 days
  expires_in: number
  openid: string
  unionid?: string
  scope: string
}

export interface WeixinToken {
  access_token: string
  refresh_token?: string
}

export interface WeixinTokenStore<T extends WeixinToken = WeixinToken> {
  /**
   * Gets token.
   */
  get(key: string): Promise<T | null>

  /**
   * Sets token.
   */
  set(key: string, token: T, ttl: number): Promise<void>

  /**
   * Delete token.
   */
  delete(key: string): Promise<void>
}

export interface WeixinAPIQuota {
  daily_limit: number
  used: number
  remain: number
}


src/core/client.ts

import { createError, type WeixinError } from './error.js'
import * as fetcher from './fetcher.js'
import type {
  WeixinAPIQuota,
  WeixinErrorResponse,
  WeixinCodeSessionResponse,
  WeixinCodeTokenResponse,
  WeixinOptions,
} from './types.js'

/**
 * Weixin Client with access token auto-retry.
 */
export class WeixinClient {
  #options: WeixinOptions

  constructor(options: WeixinOptions) {
    this.#options = options
  }

  /**
   * HTTP GET
   *
   * @throws {WeixinError}
   */
  async get<T extends object>(path: string, query: Record<string, string>): Promise<T> {
    return await this.#request('GET', path, query)
  }

  /**
   * HTTP POST
   *
   * @throws {WeixinError}
   */
  async post<T extends object>(path: string, body: Record<string, unknown>): Promise<T> {
    return await this.#request('POST', path, body)
  }

  /**
   * HTTP GET. The raw `Response` object is returned.
   *
   * @throws {WeixinError}
   */
  async getRaw(path: string, query: Record<string, string>): Promise<Response> {
    return this.#raw('GET', path, query)
  }

  /**
   * HTTP POST. The raw `Response` object is returned.
   *
   * @throws {WeixinError}
   */
  async postRaw(path: string, body: Record<string, unknown>): Promise<Response> {
    return await this.#raw('POST', path, body)
  }

  /**
   * HTTP request.
   *
   * Invalid access token error:
   * ```js
   * {
   *   errcode: 40001,
   *   errmsg: 'invalid credential, access_token is invalid or not latest, could get access_token by getStableAccessToken, more details at https://mmbizurl.cn/s/JtxxFh33r rid: 64b37c9e-5128bd63-39bd9581',
   * }
   * ```
   *
   * @throws {WeixinError}
   */
  async #request<M extends 'GET' | 'POST', T extends object>(
    method: M,
    path: string,
    params: Record<string, M extends 'GET' ? string : unknown>,
    isRetry = false,
  ): Promise<T> {
    const accessToken = await this.getAccessToken()
    const res = await fetcher.request(method, path, accessToken, params)
    const data = await res.json() as T | WeixinErrorResponse

    if (('errcode' in data && data.errcode !== 0)) {
      if (!isRetry && data.errcode === 40001) { // access token is invalid, will retry once
        await this.getAccessToken(true)
        return this.#request(method, path, params, true)
      }

      throw createError(`Request to '${path}' is failed.`, data)
    }

    return data as T
  }

  /**
   * HTTP raw request.
   *
   * @throws {WeixinError}
   */
  async #raw<M extends 'GET' | 'POST'>(
    method: M,
    path: string,
    params: Record<string, M extends 'GET' ? string : unknown>,
    isRetry = false,
  ): Promise<Response> {
    const accessToken = await this.getAccessToken()
    const res = await fetcher.request(method, path, accessToken, params)

    if (res.headers.get('content-type')?.startsWith('application/json')) { // an error arises
      const data = await res.json() as WeixinErrorResponse
      if (!isRetry && data.errcode === 40001) { // access token is invalid, will retry once
        await this.getAccessToken(true)
        return this.#raw(method, path, params, true)
      }

      throw createError(`Request to '${path}' is failed.`, data)
    }

    return res
  }

  /**
   * Gets an access token.
   *
   * @throws {WeixinError}
   */
  async getAccessToken(forceRefresh = false): Promise<string> {
    const { appid, secret, store } = this.#options
    const key = compositeKey(appid, secret.slice(0, 6), false)

    if (forceRefresh) {
      await store.delete(key)
    } else {
      const cachedToken = await store.get(key)
      if (cachedToken) return cachedToken.access_token
    }

    const data = await fetcher.createAccessToken(appid, secret)
    if ('access_token' in data) {
      const access_token = data.access_token
      const ttl = data.expires_in - 10
      await store.set(key, { access_token }, ttl)
      return access_token
    }

    throw createError('Unable to get an access token.', data)
  }

  /**
   * Gets a stable access token.
   *
   * @throws {WeixinError}
   */
  async getStableAccessToken(forceRefresh = false): Promise<string> {
    const { appid, secret, store } = this.#options
    const key = compositeKey(appid, secret.slice(0, 6), true)

    if (forceRefresh) {
      await store.delete(key)
    } else {
      const cachedToken = await store.get(key)
      if (cachedToken) return cachedToken.access_token
    }

    const data = await fetcher.createStableAccessToken(appid, secret, forceRefresh)
    if ('access_token' in data) {
      const access_token = data.access_token
      const ttl = data.expires_in - 10
      await store.set(key, { access_token }, ttl)
      return access_token
    }

    throw createError('Unable to get a stable access token.', data)
  }

  /**
   * Creates a session from an authorization code.
   *
   * @throws {WeixinError}
   */
  async createSessionFromCode(code: string): Promise<WeixinCodeSessionResponse> {
    const { appid, secret } = this.#options
    const data = await fetcher.exchangeCodeForSession(appid, secret, code)
    // What to do with `session_key`?
    if ('openid' in data)
      return data

    throw createError('Unable to create a session from an authorization code.', data)
  }

  /**
   * Creates a token from an authorization code.
   *
   * [Refresh Tokens with OAuth 2.0](https://learn.microsoft.com/en-us/linkedin/shared/authentication/programmatic-refresh-tokens)
   *
   * @throws {WeixinError}
   */
  async createTokenFromCode(code: string): Promise<WeixinCodeTokenResponse> {
    const { appid, secret } = this.#options
    const data = await fetcher.exchangeCodeForToken(appid, secret, code)
    // What to do with `access_token`, 'refresh_token'?
    if ('openid' in data)
      return data

    throw createError('Unable to create a token from an authorization code.', data)
  }

  /****** Common APIs ******/

  /**
   * Gets API quota.
   *
   * - [OpenAPI](https://developers.weixin.qq.com/doc/offiaccount/openApi/get_api_quota.html)
   *
   * @param path API path starting with '/'.
   *
   * @throws {WeixinError}
   */
  async getAPIQuota(path: string): Promise<WeixinAPIQuota> {
    const data = await this.post<{ quota: WeixinAPIQuota }>('/cgi-bin/openapi/quota/get', {
      cgi_path: path,
    })

    return data.quota
  }
}

/**
 * Gets a composite key from the given appid and secret.
 *
 * Tokens will be invalidated when `secret` is chanaged, then the `secret` is
 * involved to compute the composite key.
 */
function compositeKey(appid: string, secret: string, stable: boolean) {
  // Truncate the secret to hide details and shorten the composite key.
  return appid + ':' + secret.slice(0, 6) + (stable ? ':stable' : '')
}


src/core/memory-store.ts
import type { WeixinToken, WeixinTokenStore } from './types.js'

interface Token<T extends WeixinToken> {
  data: T
  expiresAt: number
}

/**
 * A token store in memory.
 */
export class WeixinMemoryStore<T extends WeixinToken> implements WeixinTokenStore<T> {
  tokens: Record<string, Token<T>> = Object.create(null)

  async get(key: string): Promise<T | null> {
    return new Promise<T | null>((resolve, _) => {
      setImmediate(() => {
        const data = this.#getTokenData(key)
        resolve(data)
      })
    })
  }

  async set(key: string, token: T, ttl: number): Promise<void> {
    return new Promise((resolve, _) => {
      setImmediate(() => {
        this.tokens[key] = {
          data: token,
          expiresAt: Date.now() + ttl * 1000,
        }
        resolve()
      })
    })
  }

  async delete(key: string): Promise<void> {
    return new Promise((resolve, _) => {
      setImmediate(() => {
        delete this.tokens[key]
        resolve()
      })
    })
  }

  #getTokenData(key: string): T | null {
    const token = this.tokens[key]
    if (!token) return null

    if (token.expiresAt <= Date.now()) {
      // Destroy expired token.
      delete this.tokens[key]
      return null
    }

    return token.data
  }

  /**
   * Clears all the tokens. Internal use.
   * @private
   */
  _clear() {
    this.tokens = Object.create(null)
  }
}


src/core/fetcher.ts
/**
 * HTTP GET/POST, etc.
 *
 * Put fetching in a module to allow for spying & mocking.
 *
 * Exported functions that are called by others in the same module cannot be
 * mocked. See the reason: https://stackoverflow.com/a/55193363/3414249.
 */

import type {
  WeixinAccessTokenResponse,
  WeixinErrorResponse,
  WeixinCodeSessionResponse,
  WeixinCodeTokenResponse,
} from './types.js'

const API_URL = 'https://api.weixin.qq.com'

export async function request<M extends 'GET' | 'POST'>(
  method: M,
  path: string,
  accessToken: string | null,
  params: Record<string, M extends 'GET' ? string : unknown>,
): Promise<Response> {
  if (path[0] !== '/') throw new Error("Path must start with '/'")

  if (method === 'GET') {
    const queryParams = new URLSearchParams(params as Record<string, string>)
    if (accessToken)
      queryParams.set('access_token', accessToken)

    return await fetch(API_URL + path + '?' + queryParams)
  }

  if (accessToken)
    path += `?access_token=${accessToken}`

  return await fetch(API_URL + path, {
    method,
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(params),
  })
}

/**
 * Creates an access token using OAuth2 client credentials.
 *
 * Success response:
 * ```js
 * {
 *   access_token: '70_JW9uqqofPLfob-vhwYtdxtynLfPSjKORbM0BPToN9isXQyUbsUhF4V9TPVO2QbKBbUGFO-rPEn0f_-IuD35JgEpq2VuKsem-N_w2M8vn6FQj4X4mHDLZEOKwvAQADZeAEAEHD',
 *   expires_in: 7200,
 * }
 *
 * Error response:
 * ```js
 * {
 *   errcode: 40125,
 *   errmsg: 'invalid appsecret rid: 64b79aff-477f027c-06415867',
 * }
 * ```
 *
 * - [OpenAPI](https://developers.weixin.qq.com/miniprogram/dev/OpenApiDoc/mp-access-token/getAccessToken.html)
 */
export async function createAccessToken(appid: string, secret: string): Promise<WeixinAccessTokenResponse | WeixinErrorResponse> {
  const res = await request('GET', '/cgi-bin/token', null, {
    appid,
    secret,
    grant_type: 'client_credential',
  })
  return await res.json()
}

/**
 * Creates a stable access token using OAuth2 client credentials.
 *
 * Limit:
 * - A new access token will be only generated in more than 30 seconds since
 *   the last generation.
 * - The quota is 20 times per day, if exceeded call `clear_quota` API to reset it.
 *
 * Success response:
 * ```js
 * {
 *   access_token: '70_GTSJeKUgm5yqgDjroYtToZpyOlMueufA6OU-ldrt81B--gkWgF1pgjOzCkJGqUG71DLUixGF96idq9ubfTfXGHrVO5PUrhejD0RBQHF4_acpv_iUrIWmEqc4rD4AKLaABAMWX',
 *   expires_in: 7200,
 * }
 * ```
 *
 * Quota excceded error response:
 * ```js
 * {
 *   errcode: 45009,
 *   errmsg: 'reach max api daily quota limit rid: 64b279b3-2eddb111-78605a3e'
 * }
 * ```
 *
 * - [OpenAPI](https://developers.weixin.qq.com/miniprogram/dev/OpenApiDoc/mp-access-token/getStableAccessToken.html)
 */
export async function createStableAccessToken(
  appid: string,
  secret: string,
  forceRefresh = false,
): Promise<WeixinAccessTokenResponse | WeixinErrorResponse> {
  const res = await request('POST', '/cgi-bin/stable_token', null, {
    appid,
    secret,
    grant_type: 'client_credential',
    force_refresh: forceRefresh,
  })
  return await res.json() as WeixinAccessTokenResponse | WeixinErrorResponse
}

/**
 * Exchanges an authorization code for a session.
 *
 * Used for Mini Program, Mini Game, Offical Account.
 *
 * Code must be obtained from corresponding app with the given appid.
 *
 * Success response:
 * ```js
 * {
 *   session_key: 'gPXJstHE7Gviep8pOAfvZQ==',
 *   openid: 'olQ7rLmm83trKrCNaof6Gf0IkDwU',
 * }
 *
 * Error response:
 * ```js
 * {
 *   errcode: 40029,
 *   errmsg: 'invalid code, rid: 64b79592-1db719b6-187cfe77',
 * }
 *
 * {
 *   errcode: 40163,
 *   errmsg: ‘code been used, rid: 64bb7f31-356dfe4d-6348ca3b‘,
 * },
 * ```
 *
 * - [OpenAPI](https://developers.weixin.qq.com/miniprogram/dev/OpenApiDoc/user-login/code2Session.html)
 */
export async function exchangeCodeForSession(appid: string, secret: string, code: string): Promise<WeixinCodeSessionResponse | WeixinErrorResponse> {
  const res = await request('GET', '/sns/jscode2session', null, {
    appid,
    secret,
    js_code: code,
    grant_type: 'authorization_code',
  })

  return await res.json() as WeixinCodeSessionResponse | WeixinErrorResponse
}

/**
 * Exchanges an authorization code for a token.
 *
 * Used for Weixin Open Platform, includes websites and native apps.
 *
 * Code must be obtained from corresponding app with the given appid.
 *
 * Success response:
 * ```js
 * {
 *   access_token: '70_Un4WQS_rXSSMlB5V_2av-30ELsG893aLocRvcCW21aAK9ol4T4Fjm7f3Qgcxlci2cHaSh5nOb6eMIPUd-F4EKXTfRseEUhkTQ9jHcWpkeTc',
 *   expires_in: 7200,
 *   refresh_token: '70_o_tiVeITRmgQjAiRmW0xGwkJy_0S8_0wx9LYYU4AsmGqXJFL2qA_1-QT2JvJO0RxLIEKAeE5Zq08s6boOdOflK-zVDs7cGn1W8DgyLX_GzM',
 *   openid: 'olQ7rLmm83trKrCNaof6Gf0IkDwU',
 *   scope: 'snsapi_login',
 *   unionid: 'oAg5pv3QsjETBuOSK1sUmqiYzpDc',
 * }
 *
 * Error response:
 * ```js
 * {
     errcode: 40163,
     errmsg: 'code been used, rid: 64ba6749-265e5bf9-42ca80e9',
 * }
 * ```
 *
 * - [OpenAPI](https://developers.weixin.qq.com/doc/oplatform/Website_App/WeChat_Login/Wechat_Login.html)
 */
export async function exchangeCodeForToken(appid: string, secret: string, code: string) {
  const res = await request('GET', '/sns/oauth2/access_token', null, {
    appid,
    secret,
    code,
    grant_type: 'authorization_code',
  })

  return await res.json() as WeixinCodeTokenResponse | WeixinErrorResponse
}
